crates/forge/src/lib.rs

```diff
@@ -462,26 +518,63 @@ fn run_with_fuzzing(
     };
     let mut fuzzer = RandomFuzzer::create(fuzzer_seed, fuzzer_runs, &args)?;
 
-    let mut results = vec![];
+    let mut tasks = vec![];
 
     for _ in 1..=fuzzer_runs {
         let args = fuzzer.next_args();
+        tasks.push(task::spawn({
+            let case = case.clone();
+            let package_root =  package_root.clone();
+            let runner = runner.clone();
+            let args = args.clone();
+            let cancellation_token = cancellation_token.clone();
+            let cancellation_fuzzing_token = Arc::new(token.clone());
+            let runner_params = runner_params.clone();
+            let runner_config = runner_config.clone();
+            async move {
+                tokio::select! {
+                    _ = cancellation_token.cancelled() => {
+                        // The token was cancelled
+                        Ok(TestCaseSummary::Failed { name: "fuzzing".to_string(), run_result: None, msg: None, arguments: vec![] })
+                    },
+                    _ = cancellation_fuzzing_token.cancelled() => {
+                        Ok(TestCaseSummary::Failed { name: "fuzzing".to_string(), run_result: None, msg: None, arguments: vec![] })
+                    },
+                   result = blocking_run_from_test(
+                        package_root,
+                        runner,
+                        case,
+                        runner_config,
+                        runner_params,
+                        args.clone(),
+                    ) => {
+                        result
+                    },
+                }
+            }
+        }));
+    }
 
-        let result = run_from_test_case(
-            package_root,
-            runner,
-            case,
-            runner_config.fork_targets.as_ref(),
-            &runner_params.contracts,
-            &runner_params.predeployed_contracts,
-            args.clone(),
-            &runner_params.environment_variables,
-        )?;
-        results.push(result.clone());
-
-        if let TestCaseSummary::Failed { .. } = result {
-            // Fuzz failed
-            break;
+    let mut results = vec![];
+    for task in tasks {
+        let result = match task.await? {
+            Ok(result) => result,
+            Err(e) => {
+                token.cancel();
+                return Err(e);
+            }
+        };
+
+        if let TestCaseSummary::Failed { .. } = &result {
+            if results.is_empty() {
+                token.cancel();
+                if runner_config.exit_first {
+                    cancellation_token.cancel();
+                };
+                results.push(result);
+            }
```

Comment:
I'm not sure if I understand this cancellation logic. This could be explained or preferably explained better.

Reply:
I answer in prev comment

Reply:
I wonder if this can be refactored a bit to make it more clear/readable?

Reply:
I think currently. it's better